service-base/src/main/java/fi/nls/oskari/cache/CacheManager.java

```diff
@@ -37,16 +39,27 @@
      * @return
      */
     public static <T> Cache<T> getCache(final String name) {
-        final Cache existing = CACHE_STORE.get(name);
-        if (existing != null) {
-            return existing;
-        }
+        return getCache(name, () -> new Cache<>());
+    }
 
-        // create a new one
-        final Cache<T> cache = new Cache<T>();
-        cache.setName(name);
-        CACHE_STORE.put(name, cache);
-        return cache;
+    /**
+     * Returns a cache matching name or creates one if it doesn't exist
+     * by calling the provided supplier function.
+     *
+     * @param name name of the cache
+     * @param supplier function which creates a new cache if one doesn't exists for the key
+     * @param <T>  type mapping for cache
+     * @return
+     */
+    @SuppressWarnings("unchecked")
+    public static <T> Cache<T> getCache(final String name, final Supplier<? extends Cache<T>> supplier) {
```

Comment:
Can we use something like <? extends Cache> as return value to get rid of some nasty casting when using this?

Reply:
Updated, usage example now: https://github.com/oskariorg/oskari-server/blob/33db601e0e19b4746480bf5e8c9b96274da511b9/service-wfs-client/src/main/java/org/oskari/service/wfs/client/CachingWFSClient.java#L24-L25